### ISSUE TITLE
Patch to fix out of sync subtitles

### ISSUE THREAD

**Issue by Paco8** [Sandmann79/xbmc#264]
I don't know if this problem happens in other countries as well, but least in Spain there are many movies and TV shows with the subtitles out of sync. They start correctly but gradually they are getting ahead of the video more and more. Since it's not a constant delay, the option in Kodi to adjust the delay of the subtitles is not very useful (it would be necessary to adjust it several times during the movie).

It seems the problem is that Amazon is incorrectly providing subtitles with times for 24 fps while the videos actually play at 23.976 fps. Converting the times from 24 fps to 23.976 fps fixes the subtitles. Unfortunately I couldn't find an option in Kodi to change the fps of subtitles on the fly, so I made a small modification in the Amazon VOD addon. 

It adds the option "Provide 23.976 fps subtitles" in the addon preferences. When enabled, it creates a copy of the subtitles but with the times converted from 24 to 23.976 fps. So if you start to watch a movie and after some minutes you realize that the subtitles are getting ahead of the video, just select the 23.976 version of the subtitles and they'll be fine for the rest of the movie.

[change-subtitle-fps.zip](https://github.com/Sandmann79/xbmc/files/2882267/change-subtitle-fps.zip)

**Comment by Varstahl** [comment]
What you're saying doesn't really make much sense though, the subtitles are frame agnostic, they're time based. 2 seconds at 24 fps is the same timestamp as 2 seconds in 23.976 fps. Are you saying that some subtitles have incorrect timestamps? I've never noticed anything like it. Are you sure it's just not Kodi acting up? That sounds rather weird.

**Comment by Paco8** [comment]
Maybe the term is not correct but the fact is that when this problem happens the subtitles play a little bit faster than the video. It's not a problem of kodi, it happens also with the primevideo official applications. Yes, the timestamps are wrong. At least is Spain this problem happens in a lot of movies and TV shows.
An example is the movie "Pitch Perfect".

**Comment by Varstahl** [comment]
I'm working on a proxy system that will enhance things, I guess I can slip in a time remapper.

**Comment by Varstahl** [comment]
Initial version implemented in 6da25ed7ee22bebe8a436b63c26652b98398203a

**Comment by Paco8** [comment]
Wonderful. I've just tested it and it works fine. Thank you very much.

**Comment by Paco8** [comment]
I watched a movie a couple of days ago, again with the subtitles out of sync. However this time instead of getting ahead, the subtitles were gradually getting behind of the video. In order to fix the synchronization I had to convert from 23.976 to 24 fps.

So I thought it would be good if it was possible to configure the stretching factor in the addon. I attach a patch.

[stretch_factor.zip](https://github.com/Sandmann79/xbmc/files/2959071/stretch_factor.zip)
